File: lc/coreprep/implementation/patterns_reference_implementation/k_way_merge/smallest_number_range.cpp
#include <algorithm>
#include <iostream>
#include <limits>
#include <queue>
#include <vector>

using namespace std;

struct Entry {
  int list_idx;
  int arr_idx;

  Entry(int list_idx, int arr_idx) : list_idx(list_idx), arr_idx(arr_idx) {}
};

struct CompareEntry {
  const vector<vector<int>> &lists;

  CompareEntry(const vector<vector<int>> &lists) : lists(lists) {}

  bool operator()(const Entry &e1, const Entry &e2) {
    return lists[e1.list_idx][e1.arr_idx] > lists[e2.list_idx][e2.arr_idx];
  }
};

// O(nlogm) / O(m)
pair<int, int> smallest_number_range(const vector<vector<int>> &lists) {
  CompareEntry comp(lists);
  priority_queue<Entry, vector<Entry>, CompareEntry> min_heap(comp);

  int curr_start = 0, curr_end = numeric_limits<int>::max();
  int curr_max = numeric_limits<int>::min();
  for (int i = 0; i < lists.size(); i++) {
    min_heap.push(Entry(i, 0));
    curr_max = max(curr_max, lists[i][0]);
  }

  while (min_heap.size() == lists.size()) {
    Entry entry = min_heap.top();
    min_heap.pop();

    if (curr_max - lists[entry.list_idx][entry.arr_idx] <
        curr_end - curr_start) {
      curr_start = lists[entry.list_idx][entry.arr_idx];
      curr_end = curr_max;
    }

    entry.arr_idx++;

    if (entry.arr_idx < lists[entry.list_idx].size()) {
      min_heap.push(entry);
      curr_max = max(curr_max, lists[entry.list_idx][entry.arr_idx]);
    }
  }

  return make_pair<>(curr_start, curr_end);
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, m, n;
  cin >> tt;

  while (tt--) {
    cin >> m;

    vector<vector<int>> lists;
    for (int i = 0; i < m; i++) {
      cin >> n;

      vector<int> a(n);
      for (int j = 0; j < n; j++) {
        cin >> a[j];
      }
      lists.emplace_back(a);
    }

    auto [start, end] = smallest_number_range(lists);
    cout << start << " " << end << "\n";
  }

  return 0;
}

File: lc/coreprep/implementation/patterns_reference_implementation/k_way_merge/SmallestNumberRange.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class SmallestNumberRange {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int m = in.nextInt();

        List<Integer[]> lists = new ArrayList<>();
        for (int i = 0; i < m; i++) {
          int n = in.nextInt();

          Integer[] a = new Integer[n];
          for (int j = 0; j < n; j++) {
            a[j] = in.nextInt();
          }
          lists.add(a);
        }

        Range minRange = smallestNumberRange(lists);
        System.out.printf("%d %d\n", minRange.start(), minRange.end());
      }
    }
  }

  record Range(int start, int end) {}

  static class Entry {
    int listIndex;
    int arrIndex;

    Entry(int listIndex, int arrIndex) {
      this.listIndex = listIndex;
      this.arrIndex = arrIndex;
    }
  }

  // O(nlogm) / O(m)
  public static Range smallestNumberRange(List<Integer[]> lists) {
    PriorityQueue<Entry> minHeap = new PriorityQueue<>(
        (Entry e1, Entry e2)
            -> Integer.compare(lists.get(e1.listIndex)[e1.arrIndex],
                               lists.get(e2.listIndex)[e2.arrIndex]));

    int currStart = Integer.MIN_VALUE0, currEnd = Integer.MAX_VALUE;
    int currMax = Integer.MIN_VALUE;
    for (int i = 0; i < lists.size(); i++) {
      minHeap.add(new Entry(i, 0));
      currMax = Math.max(currMax, lists.get(i)[0]);
    }

    while (minHeap.size() == lists.size()) {
      Entry entry = minHeap.poll();

      if (currMax - lists.get(entry.listIndex)[entry.arrIndex] <
          currEnd - currStart) {
        currStart = lists.get(entry.listIndex)[entry.arrIndex];
        currEnd = currMax;
      }

      entry.arrIndex++;

      if (entry.arrIndex < lists.get(entry.listIndex).length) {
        minHeap.add(entry);
        currMax = Math.max(currMax, lists.get(entry.listIndex)[entry.arrIndex]);
      }
    }

    return new Range(currStart, currEnd);
  }
}